### PR TITLE
Vrfv2 polygonedge support

### PR DESCRIPTION
**Pull request by kidambisrinivas**
## VRF Issue with Nexon chain

- Block hashes on Polygon Edge are not same as regular EVM chains.
- Nexon chain has the property: keccak256(RLPencoding(blockheader)) == blockhash, but performs RLP encoding in a slightly different manner than other EVM chains
  - Nexon chain removes blockHeader.ExtraData.Committed field before RLP encoding the block header
  - Minor variations for Nonce field (empty nonce represented as 0x0000000000000000)
  - baseFeePerGas JSON field present as baseFee in Nexon blockchain response. [Seems to have been fixed in polygon-edge framework already, but Nexon chain, still seems to send baseFee, had to handle that]
- **Store BHS backwards functionality** of VRF V2 does not work: **Current PR fixes this**

## Solutions

- **Approach1**: Use [polygon-edge](github.com/0xPolygon/polygon-edge) to RLP encode block headers
  - Adding polygon-edge as a dependency caused a lot of issues with conflicting dependency version with other libraries in this repo and some methods being referenced from older versions.
- **Approach2**: Replicate RLP encoding logic of polygon-edge package. [Current PR]

### CODE REVIEW

File: core/scripts/common/helpers.go [hunk 1]
```diff
@@ -452,3 +483,38 @@ func IsAvaxNetwork(chainID int64) bool {
...
+func CalculateLatestBlockHeader(env Environment, blockNumberInput int) (err error) {
+	blockNumber := uint64(blockNumberInput)
+	if blockNumberInput == -1 {
+		blockNumber, err = env.Ec.BlockNumber(context.Background())
+		if err != nil {
+			return fmt.Errorf("failed to fetch latest block: %+v", err)
+		}
+	}
+
+	blockNumber = blockNumber - 1
```
Comment (jinhoonbang) [inline core/scripts/common/helpers.go:502]: do we have to subtract 1 here? maybe a comment would be helpful

Comment (jinhoonbang) [reply core/scripts/common/helpers.go:502]: oh I see that It validates blockhash below

File: core/scripts/common/polygonedge.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	dst = val.MarshalTo(dst)
+
+	return dst, h.Hash.String(), err
+}
+
+// MarshalRLPWith marshals the header to RLP with a specific fastrlp.Arena
+// Adding polygon-edge as a dependency caused a lot of issues with conflicting
+// dependency version with other libraries in this repo and some methods being
+// referenced from older versions
+// Reference: https://github.com/0xPolygon/polygon-edge/blob/develop/types/rlp_marshal.go#L73C50-L73C53
```
Comment (jinhoonbang) [inline core/scripts/common/polygonedge.go:92]: I wonder if Polygon can change this implementation in the future. How bad were the dependency issues?

File: core/scripts/common/hex.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+func DecodeString(str string) ([]byte, error) {
+	return hex.DecodeString(str)
+}
+
+// DecodeHex converts a hex string to a byte array
+func DecodeHex(str string) ([]byte, error) {
+	str = strings.TrimPrefix(str, "0x")
+
+	return hex.DecodeString(str)
+}
```
Comment (makramkd) [inline core/scripts/common/hex.go:28]: Making wrapper methods like this should be avoided. Use the standard library directly, and for the `0x` prefixed hexes, use `hexutil.Encode` from go-ethereum.

File: core/scripts/common/polygonedge.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	GasUsed      hexutil.Uint64 `json:"gasUsed"          gencodec:"required"`
+	Timestamp    hexutil.Uint64 `json:"timestamp"        gencodec:"required"`
+	ExtraData    ExtraData      `json:"extraData"        gencodec:"required"`
+	MixHash      common.Hash    `json:"mixHash"`
+	Nonce        Nonce          `json:"nonce"`
+	Hash         common.Hash    `json:"hash"`
+
+	// BaseFee was added by EIP-1559 and is ignored in legacy headers.
+	BaseFee    hexutil.Uint64 `json:"baseFeePerGas"`
+	BaseFeeAlt hexutil.Uint64 `json:"baseFee,omitempty"`
```
Comment (makramkd) [inline core/scripts/common/polygonedge.go:65]: Can you document this and why its there in addition to `BaseFee`?

File: core/scripts/vrfv2plus/testnet/main.go [hunk 5]
```diff
@@ -1026,21 +1026,14 @@ func main() {
...
-		wrapperAddress := cmd.String("wrapper-address", "", "address of the VRFV2Wrapper contract")
-		wrapperGasOverhead := cmd.Uint("wrapper-gas-overhead", 50_000, "amount of gas overhead in wrapper fulfillment")
-		coordinatorGasOverhead := cmd.Uint("coordinator-gas-overhead", 52_000, "amount of gas overhead in coordinator fulfillment")
-		wrapperPremiumPercentage := cmd.Uint("wrapper-premium-percentage", 25, "gas premium charged by wrapper")
-		keyHash := cmd.String("key-hash", "", "the keyhash that wrapper requests should use")
-		maxNumWords := cmd.Uint("max-num-words", 10, "the keyhash that wrapper requests should use")
-		helpers.ParseArgs(cmd, os.Args[2:], "wrapper-address", "key-hash")
+		wrapperAddress := cmd.String("wrapper-address", "", "address of the VRFV2PlusWrapper contract")
+		coordinatorAddress := cmd.String("coordinator-address", "", "address of the VRFV2Plus coordinator contract")
+		subID := cmd.Uint("sub-id", 1, "subscription ID")
```
Comment (makramkd) [inline core/scripts/vrfv2plus/testnet/main.go:1031]: Why the changes here?

File: core/scripts/vrfv2plus/testnet/super_scripts.go [hunk 6]
```diff
@@ -231,11 +231,6 @@ func deployWrapperUniverse(e helpers.Environment) {
 	linkAddress := cmd.String("link-address", "", "address of link token")
 	linkETHFeedAddress := cmd.String("link-eth-feed", "", "address of link-eth-feed")
 	coordinatorAddress := cmd.String("coordinator-address", "", "address of the vrf coordinator v2 contract")
-	wrapperGasOverhead := cmd.Uint("wrapper-gas-overhead", 50_000, "amount of gas overhead in wrapper fulfillment")
-	coordinatorGasOverhead := cmd.Uint("coordinator-gas-overhead", 52_000, "amount of gas overhead in coordinator fulfillment")
-	wrapperPremiumPercentage := cmd.Uint("wrapper-premium-percentage", 25, "gas premium charged by wrapper")
-	keyHash := cmd.String("key-hash", "", "the keyhash that wrapper requests should use")
-	maxNumWords := cmd.Uint("max-num-words", 10, "the keyhash that wrapper requests should use")
```
Comment (makramkd) [inline core/scripts/vrfv2plus/testnet/super_scripts.go:238]: Reasoning behind this?

File: core/scripts/common/polygonedge.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to get poloygon-edge header: %+v", err)
+	}
+
+	ar := &fastrlp.Arena{}
+	val, err := MarshalRLPWith(ar, &h)
+	if err != nil {
+		return nil, "", err
+	}
+	// fastrlp.DefaultArenaPool.Put(ar)
```
Comment (makramkd) [inline core/scripts/common/polygonedge.go:80]: Can delete.

File: core/scripts/common/helpers.go [hunk 8]
```diff
@@ -396,9 +403,13 @@ func BinarySearch(top, bottom *big.Int, test func(amount *big.Int) bool) *big.In
 // Get RLP encoded headers of a list of block numbers
 // Makes RPC network call eth_getBlockByNumber to blockchain RPC node
 // to fetch header info
-func GetRlpHeaders(env Environment, blockNumbers []*big.Int) (headers [][]byte, err error) {
+func GetRlpHeaders(env Environment, blockNumbers []*big.Int) (headers [][]byte, hashes []string, err error) {
+
+	hashes = make([]string, len(blockNumbers))
```
Comment (makramkd) [inline core/scripts/common/helpers.go:408]: Can avoid this and just use `append` for more readable code.

File: core/scripts/common/helpers.go [hunk 9]
```diff
@@ -408,39 +419,59 @@ func GetRlpHeaders(env Environment, blockNumbers []*big.Int) (headers [][]byte,
...
 			// We can still use vanilla go-ethereum rlp.EncodeToBytes, see e.g
 			// https://github.com/ava-labs/coreth/blob/e3ca41bf5295a9a7ca1aeaf29d541fcbb94f79b1/core/types/hashing.go#L49-L57.
-			rlpHeader, err := rlp.EncodeToBytes(h)
+			rlpHeader, err = rlp.EncodeToBytes(h)
 			if err != nil {
-				return nil, fmt.Errorf("failed to encode rlp: %+v", err)
+				return nil, hashes, fmt.Errorf("failed to encode rlp: %+v", err)
 			}
 
+			if hashes != nil && len(hashes) > i {
```
Comment (makramkd) [inline core/scripts/common/helpers.go:431]: Nil check is not necessary here, can just do `hashes = append(hashes, h.Hash())`

**Review by makramkd — COMMENTED** [review]
Looking good! Some requested changes below

File: core/scripts/common/helpers.go (hunk 1 above)
Comment (kidambisrinivas) [reply core/scripts/common/helpers.go:502]: GetRLPHeaders method **increments** the blockNum sent by 1 and then fetches block headers for the child block. So, in order to get the block header for a given block, I had to decrement by 1. 

This is to give a user-friendly command like: `go run ./vrfv2plus/testnet latest-block-header --block-number 453551`

File: core/scripts/common/polygonedge.go (hunk 2 above)
Comment (kidambisrinivas) [reply core/scripts/common/polygonedge.go:92]: The dependency issues seemed like a rabbit hole. After fixing the dependency version issues, I ran into some undefined methods problem, probably due to methods being referenced from older versions. 

We should consider relying on polygon-edge in future, when the dependency issues resolve.

File: core/scripts/common/helpers.go [hunk 10]
```diff
@@ -31,6 +32,8 @@ type Environment struct {
 	Owner *bind.TransactOpts
 	Ec    *ethclient.Client
 
+	PolygonEdgeClient *rpc.Client
```
[the remark this answers is not in the document]
Comment (kidambisrinivas) [reply core/scripts/common/helpers.go:35]: Sure

[the remark this answers is not in the document]
Comment (kidambisrinivas) [reply core/scripts/common/helpers.go:35]: Done

File: core/scripts/common/helpers.go [hunk 11]
```diff
@@ -396,9 +403,13 @@ func BinarySearch(top, bottom *big.Int, test func(amount *big.Int) bool) *big.In
...
 // Makes RPC network call eth_getBlockByNumber to blockchain RPC node
 // to fetch header info
-func GetRlpHeaders(env Environment, blockNumbers []*big.Int) (headers [][]byte, err error) {
+func GetRlpHeaders(env Environment, blockNumbers []*big.Int) (headers [][]byte, hashes []string, err error) {
+
+	hashes = make([]string, len(blockNumbers))
+
 	headers = [][]byte{}
-	for _, blockNum := range blockNumbers {
+	var rlpHeader []byte
```
[the remark this answers is not in the document]
Comment (kidambisrinivas) [reply core/scripts/common/helpers.go:411]: Ah ok, I felt this line was being replicated in all of conditions of the if statement, so pulled it out. As we've more cases in future, this may be replicated more.

`headers = append(headers, rlpHeader)`

Do let me know if you'd like to keep it this way or remove the forward declaration approach

File: core/scripts/common/hex.go (hunk 3 above)
Comment (kidambisrinivas) [reply core/scripts/common/hex.go:28]: Yes, sure. Fixed this.

File: core/scripts/common/polygonedge.go (hunk 4 above)
Comment (kidambisrinivas) [reply core/scripts/common/polygonedge.go:65]: Sure

File: core/scripts/vrfv2plus/testnet/super_scripts.go (hunk 6 above)
Comment (kidambisrinivas) [reply core/scripts/vrfv2plus/testnet/super_scripts.go:238]: This change seems to have been missed out in VRF V2+ Super Script PR and actually seem to break VRF V2+ commands

File: core/scripts/common/polygonedge.go (hunk 7 above)
Comment (kidambisrinivas) [reply core/scripts/common/polygonedge.go:80]: Sure

File: core/scripts/common/helpers.go (hunk 9 above)
Comment (kidambisrinivas) [reply core/scripts/common/helpers.go:431]: Yep, I'd originally sent in hashes instead of returning it. Will fix it now

File: core/scripts/common/polygonedge.go [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+// github.com/0xPolygon/polygon-edge. polygon-edge is a node implementation,
+// and not a client. Adding polygon-edge as a dependency caused a lot of
+// issues with conflicting dependency version with other libraries in this
+// repo and some methods being referenced from older versions.
+func GetIbftExtraClean(extra []byte) (cleanedExtra []byte, err error) {
+	// Capture prefix 0's sent by nexon supernet
+	hexExtra := hex.EncodeToString(extra)
+	prefix := ""
+	for _, s := range hexExtra {
+		if string(s) == string("0") {
```
Comment (makramkd) [inline core/scripts/common/polygonedge.go:152]: You can compare runes:
```suggestion
		if s == '0' {
```

File: core/scripts/vrfv2plus/testnet/main.go (hunk 5 above)
Comment (makramkd) [reply core/scripts/vrfv2plus/testnet/main.go:1031]: Not sure I understand, why was this broken?

**Review by bolekk — CHANGES_REQUESTED** [review]
Please rebase this PR properly. It now touches a lot of unrelated files with incorrect changes.

**Comment by kidambisrinivas** [comment]
Rebase went wrong and added many commits to this PR. Closing this
